mail_debrand/models/mail_render_mixin.py

```diff
@@ -12,43 +12,61 @@
+                    part
+                )  # html with broken links   tree = etree.fromstring(value) just xml
+                if tree is None:
+                    new_parts.append(part)
+                    continue
+                odoo_achors = tree.xpath('//a[contains(@href,"odoo.com")]')
+                for elem in odoo_achors:
+                    parent = elem.getparent()
+                    previous = elem.getprevious()
 
-                if remove_before and not remove_parent and previous is not None:
-                    # remove 'using' that is before <a and after </span>
-                    bytes_text = etree.tostring(
-                        previous, pretty_print=True, method="html"
-                    )
-                    only_what_is_in_tags = bytes_text[: bytes_text.rfind(b">") + 1]
-                    data_formatted = html.fromstring(only_what_is_in_tags)
-                    parent.replace(previous, data_formatted)
-                if remove_parent and len(parent.getparent()):
-                    # anchor <a href odoo has a parent powered by that must be removed
-                    parent.getparent().remove(parent)
-                else:
-                    if parent.tag == "td":  # also here can be powerd by
+                    if remove_before and not remove_parent and previous is not None:
+                        # remove 'using' that is before <a and after </span>
+                        bytes_text = etree.tostring(
+                            previous, pretty_print=True, method="html"
+                        )
+                        only_what_is_in_tags = bytes_text[: bytes_text.rfind(b">") + 1]
+                        data_formatted = html.fromstring(only_what_is_in_tags)
+                        parent.replace(previous, data_formatted)
+                    if remove_parent and len(parent.getparent()):
+                        # anchor <a href odoo has a parent powered by that must be removed
                         parent.getparent().remove(parent)
                     else:
-                        parent.remove(elem)
-            value = etree.tostring(tree, pretty_print=True, method="html")
-            # etree can return bytes; ensure we get a proper string
-            if type(value) is bytes:
-                value = value.decode()
-        return re.sub("[^(<)(</)]odoo", "", value, flags=re.IGNORECASE)
+                        if parent.tag == "td":  # also here can be powerd by
+                            parent.getparent().remove(parent)
+                        else:
+                            parent.remove(elem)
+                part = etree.tostring(tree, pretty_print=True, method="html")
+                # etree can return bytes; ensure we get a proper string
+                if type(part) is bytes:
+                    part = part.decode()
+                new_parts.append(
+                    re.sub("[^(<)(</)]odoo", "", part, flags=re.IGNORECASE)
```

Comment:
Are there really any `<odoo> whatever </odoo>` elements added to emails? 🤔

Reply:
Well, I think this regex is actually badly constructed, as @ryanc-me also suggested and tried to fix in https://github.com/OCA/social/pull/714

Right now, it is actually matching (according to https://regexr.com):
- A negated set of:
    - a `(` character
    - a `<` character
    - a `)` character
    - all the above again :sweat_smile: 
    - a `/` character
- The `odoo` word

So, it obviously needs changing... However, as @ryanc-me seems to have a fix ready in the PR mentioned above, I just wanted to fix another part of the problem, and then maybe we can proceed from what he already did...

Reply:
@joao-p-marques you're exactly right about the regex, there's a bit more info in #713. The PR #714 does "fix" that issue (see this [regex101](https://regex101.com/r/o01Rfz/5)), but I don't think the regex approach is actually correct. It leaves us with `<span>Powered by <a></a></span>`, which is *technically* de-branding, but isn't a nice solution.

I'm not convinced we can remove all branding with a generic regex - maybe instead we need to find all strings like "Odoo" in the  CE/EE sources, and write some specific rules?


Also @Yajo, just for reference, I think the original author was trying not to match the `<odoo></odoo>` tags surrounding all Odoo templates (the top-level element in all XML docs)

Reply:
Thanks for the explanation!

Well, with this PR, as we are making sure to leave the message body intact, we can be a bit more "strict" while reformatting the rest, which already solves a big part of the problem... From my testing, even with this wrong regex, we are being able to actually hide what we want without removing too much, so that takes some urgency out of the other part.

Once this is merged, maybe you can rebase your PR so that we can keep the discussion there and try out your changes.

Reply:
> From my testing, even with this wrong regex, we are being able to actually hide what we want

Yes, that's what the code tells me. AFAICS we're already detecting the "odoo" parts properly and removing the garbage. This regex is useless now (it could be before too, but now it certainly is), so please just remove it.

In case there's some corner case not being detected now, then a future PR can fix it and provide a valid test for it.

Reply:
Ok, removed then. Let's work on reviewing and #714 to make sure we don't miss anything and don't introduce any other errors...